File: LeetCode/Tree/TraverseTree/Leetcode_508_most_frequent_subtree_sum/Leetcode_508_most_frequent_subtree_sum.cpp
/**
 * @brief: Leetcode_508_出现次数最多的子树元素和
 * @link: https://leetcode.cn/problems/most-frequent-subtree-sum/description/
 * @author: liangj.zhang
 * @date: 21/11/2025
 * 
 * @Difficulty: Medium
 * 
 * @Label: Tree & HashMap
 * 
 * @Retrospect(worthy 1 - 5): 3
 * 
 * @thoughts:
 *  值得学习使用 std::max_element，以及比较器的两种写法；
 * 
 *      + 分析：
 *          + 时间复杂度：遍历树 O(n) + max_element() 查找最大值 O(n) + 再遍历找 key O(n)
 *          + 空间复杂度：O(h) + O(n) => O(n)
 *      + 写法 1
 *          + rank:
 *              + 时间效率：8 ms, 击败 47.68%
 *              + 空间效率：24.27 MB, 击败 95.94%
 *      + 写法 2
 *          + rank:
 *              + 时间效率：3 ms, 击败 99.03%
 *              + 空间效率：24.32 MB, 击败 94.79%
 */

#include "../../../Include/Leetcode/Tree/Tree.h"
using Leetcode::Tree::BinaryTree::TreeNode;

#include <algorithm>
#include <vector>
using std::vector;

#include <unordered_map>
using std::unordered_map;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#define um unordered_map

class Solution {
private:
    static um<int, int> rec;
    static bool Cmp(const std::pair<const int, int>& a,
                    const std::pair<const int, int>& b) {
        return a.second < b.second;
    }

    int __findFrequentTreeSum(TreeNode* root) {
        
        if (!root) return 0;

        int left = __findFrequentTreeSum(root->left);
        int right = __findFrequentTreeSum(root->right);

        int sum = left + right + root->val;
        if (rec.find(sum) != rec.end()) 
            ++rec[sum];
        else
            rec[sum] = 1;

        return sum;
    }
public:
    // 写法 2
    vector<int> findFrequentTreeSum(TreeNode* root) {
        
        rec.clear();
        vector<int> ret;
        __findFrequentTreeSum(root);
        int maxValue = std::max_element(rec.begin(), rec.end(), Cmp)->second;
        for (const auto& [k, v] : rec) {

            if (v == maxValue) {
                ret.push_back(k);
            }
        }

        return ret;
    }

    // 写法 1
    // vector<int> findFrequentTreeSum(TreeNode* root) {
        
    //     rec.clear();
    //     vector<int> ret;
    //     __findFrequentTreeSum(root);
    //     int maxValue = std::max_element(rec.begin(), rec.end(), [](const auto& a, const auto& b) {
    //                         return a.second < b.second;
    //     })->second;
    //     for (const auto& [k, v] : rec) {

    //         if (v == maxValue) {
    //             ret.push_back(k);
    //         }
    //     }

    //     return ret;
    // }
};

um<int, int> Solution::rec{};

File: LeetCode_Python/Tree/TraverseTree/Leetcode_508_most_frequent_subtree_sum/Leetcode_508_most_frequent_subtree_sum.py
"""
@brief: Leetcode_508_出现次数最多的子树元素和
@link: https://leetcode.cn/problems/most-frequent-subtree-sum/description/
@author: liangj.zhang
@date: 24/11/2025

@Difficulty: Medium

@Label: Tree & HashMap

@Retrospect(worthy 1 - 5): 3

@thoughts:
  值得学习使用 max() / sorted() 的自定义比较器，以及对应的两种写法；
    ps: python 写法值的回顾的地方：获得 dict 的 values()，然后得到最大值；列表推导式

  + 分析：
      + 时间复杂度：遍历树 O(n) + 查找最大频次 O(n) + 再遍历找 key O(n)
      + 空间复杂度：O(h) + O(n) => O(n)

  + 写法 1
      + rank:
          + 时间效率：8 ms, 击败 47.68%
          + 空间效率：24.27 MB, 击败 95.94%

  + 写法 2
      + rank:
          + 时间效率：3 ms, 击败 99.03%
          + 空间效率：24.32 MB, 击败 94.79%
"""

import os
import sys

print(os.path.join(os.path.dirname(__file__)))
print(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../.."))) #f:\Code\zljgit\LeetCode
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../..")))

from typing import List, Optional

from Include.PyLeetcode.Tree.Tree import TreeNode

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self) -> None:
        self.rec = {}

    def __findFrequentTreeSum(self, root: Optional[TreeNode]) -> int:
        if not root: return 0
        left = self.__findFrequentTreeSum(root.left)
        right = self.__findFrequentTreeSum(root.right)
        _sum = left + right + root.val
        if self.rec.get(_sum) is not None:
            self.rec[_sum] += 1
        else:
            self.rec[_sum] = 1
        return _sum

    def findFrequentTreeSum(self, root: Optional[TreeNode]) -> List[int]:
        self.__findFrequentTreeSum(root)
        _max = max(self.rec.values())
        # ret = []
        # for k, v in self.rec.items():
        #     if v == _max:
        #         ret.append(k)
        return [k for k, v in self.rec.items() if v == _max]
    